### PR TITLE
Experiment: A Modules API with optional static server dependency graph

### PR DESCRIPTION
**Pull request by luisherranz**
## What?

Built on top of https://github.com/WordPress/gutenberg/pull/56092.

This is the second of a series of experiments to understand the requirements of a hypothetical Modules API. More info in the Tracking Issue:

- https://github.com/WordPress/gutenberg/issues/55942

This experiment aims to implement an API that is slightly more complex than [the one implemented in the first experiment](https://github.com/WordPress/gutenberg/pull/56092), with the same characteristics but adding an optional static server dependency graph to be able to preload modules and avoid initial waterfalls:

- An API to register modules using:
  - A module identifier (required)
  - The full URL or the path (required)
  - Whether the module is for the admin, the frontend or both (required)
  - A version (optional)
  - The static dependencies (optional)
- An API to enqueue modules.
- All registered modules add an entry in the import map.
- Enqueued modules add a `<script>` tag in the head.
- The static dependencies are preloaded when the module is enqueued using a `<link rel="modulepreload">` tag in the head.

## Why?

In the [first experiment](https://github.com/WordPress/gutenberg/pull/56092), it's not easy to know which modules need to be preloaded and, therefore, it's not easy to avoid the initial waterfalls.

## How?

By leveraging a new `Gutenberg_Modules` class that acts as a singleton and a couple of functions to expose the functionality:

- `gutenberg_register_module` to register modules and add entries to the import map.
- `gutenberg_enqueue_module` to enqueue modules.

Then, we print three times using the `wp_head` action:

- The import map.
- The enqueued modules.
- The preloaded modules.

## Testing Instructions

- Add some interactive blocks: Navigation, Query or Image block.
- Check that in the frontend, there is an import map script.
- Check that scripts with `type="module"` are loaded by the interactive blocks.
- Check that the `@wordpress/interactivity` module is loaded by the other modules using `import ... from "@wordpress/interactivity"`.
- Check that the `@wordpress/interactivity` module is download at the same time than the rest (instead of after, like in the [first experiment](https://github.com/WordPress/gutenberg/pull/56092)), and that there is a link tag with `rel="modulepreload"` in the head.

### CODE REVIEW

File: tools/webpack/interactivity.js [hunk 1]
```diff
@@ -12,18 +12,28 @@ module.exports = {
...
 		filename: './build/interactivity/[name].min.js',
+		library: {
+			type: 'module',
+		},
 		path: join( __dirname, '..', '..' ),
+		environment: { module: true },
+	},
+	externalsType: 'module',
+	externals: {
+		'@wordpress/interactivity': '@wordpress/interactivity',
```
Comment (youknowriad) [inline tools/webpack/interactivity.js:36]: I think at some point in our config we need to maintain an esmodules array to keep these as is in the code (just like we have a dedicated config for bundled modules). I expect interactivity to be just one of these modules

File: packages/block-library/src/search/index.php [hunk 2]
```diff
@@ -203,27 +191,18 @@ function register_block_core_search() {
...
+		'frontend',
+		array(
+			'version'      => defined( 'GUTENBERG_VERSION' ) ? GUTENBERG_VERSION : get_bloginfo( 'version' ),
+			'dependencies' => array( '@wordpress/interactivity' ),
+		)
+	);
 }
-
-add_action( 'wp_print_scripts', 'block_core_search_ensure_interactivity_dependency' );
+add_action( 'init', 'register_block_core_search' );
```
Comment (youknowriad) [inline packages/block-library/src/search/index.php:205]: I really like how this simplifies the code of all these "interactive" blocks.

File: tools/webpack/interactivity.js (hunk 1 above)
Comment (luisherranz) [reply tools/webpack/interactivity.js:36]: Yeah. I didn't pay much attention to the bundling side yet as it shouldn't affect the PHP API, so I just made it work 😄

_(I'm not a fan of the code produced by Webpack for ES modules, by the way.)_

**Comment by gziolo** [comment]
This is now part of WordPress core 🎉

The final approach is documented in the dev note: [Script Modules in 6.5](https://make.wordpress.org/core/2024/03/04/script-modules-in-6-5/).

**Pull request closed without merging**